File: 0733-flood-fill/0733-flood-fill.cpp
class Solution {
public:
    vector<vector<int>> floodFill(vector<vector<int>>& image, int sr, int sc, int color) {
        int initial_color = image[sr][sc];
        queue<pair<int, int>> q;
        q.push({sr, sc});
        vector<vector<int>> new_image = image;
        new_image[sr][sc] = color;
        int n = image.size();
        int m = image[0].size();

        int drow[]  = {-1, 0, 1, 0};
        int dcol[] = {0, 1, 0, -1};

        while(!q.empty()){
            int row = q.front().first;
            int col = q.front().second;
            q.pop();

            for(int i = 0; i < 4; i++){
                int nrow = row + drow[i];
                int ncol = col + dcol[i];

                if(nrow >=0 && nrow < n && ncol >=0 && ncol < m && new_image[nrow][ncol] != color && image[nrow][ncol] == initial_color){
                    new_image[nrow][ncol] = color;
                    q.push({nrow, ncol});
                }
            }
        }
        return new_image;
    }
};

File: 0733-flood-fill/0733-flood-fill.py
class Solution:
    def floodFill(self, image: List[List[int]], sr: int, sc: int, color: int) -> List[List[int]]:
        key = image[sr][sc]
        image[sr][sc] = color
        n = len(image)
        m = len(image[0])
        vis = [[0] * m for _ in range(n)]
        vis[sr][sc] = 1
        drow = [-1, 0, 1, 0]
        dcol = [0, 1, 0, -1]
        q = deque()
        q.append((sr, sc))

        while q:
            (row, col) = q.popleft()

            for i in range(4):
                nrow = row + drow[i]
                ncol = col + dcol[i]

                if nrow >= 0 and nrow < n and ncol >= 0 and ncol < m and image[nrow][ncol] == key and vis[nrow][ncol] == 0:
                    vis[nrow][ncol] = 1
                    image[nrow][ncol] = color
                    q.append((nrow, ncol))
        
        return image
